src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -607,6 +616,9 @@ private static boolean supportsPortFileName(String war) throws IOException {
     }
 
     private static int readPort(File portFile) throws IOException {
+        if (!portFile.canRead()) {
```

Comment:
It’s just my assumption, because I still can’t reproduce the failure, but perhaps something wrong with the file read, could it be the case when file read happens at the same time when port value written and we receive IOException? (reading the logs more carefully I see that when flakes happens the test overall time about 2-3 seconds (usual time 15-20 sec) it means that the Jenkins is just above the start and it’s almost immediately stops)

Reply:
We could remove this log before merge, I just want somehow to run failing tests on a ci using my fork and examine the logs then.

Reply:
I managed to reproduce locally flaky failure and the reason is because port value is still empty we catch NumberFormatException and throw AssertionError (it was lost before and now visible because of [61659b8](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/61659b8e909e2bd3ca76fa75a50122e6e91b6ea3)), what are your thoughts if we retry to read it until port gets populated? see my idea in [8106c20](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/8106c20e0d8464e49a734fbb082a79411cab5d28).

Reply:
well that is an interesting finding :)

https://github.com/jenkinsci/winstone/blob/4754da67bf797b906a5c858b2de63b97ac609543/src/main/java/winstone/Launcher.java#L232-L234 is indeed a source of potential flakyness (that may or may not be your flake) but it looks like it could be.

The file and its contents are not created atomically - so if you get very unlucky the file will exist before it is being written to.
The upstream could be changed (maybe it should) to use an atomic write (write and flush to a intermediate file, then move the file in place with an atomic flag set, but this may really be overkill).  if you want to fix the flakes in the interim then probably if the port number is not valid sleep 5 seconds and try again once only.

Reply:
IMO the right fix would be [49689e7](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/49689e76453cdda1940301f0c2b96490e032358d) . It is the while loop which is waiting until Jenkins complete the start procedure, so the condition when file is ready to provide the port value is not fully correct. (I'm going to test my changes and run flaky tests with this changes to see if they will be still flaky).

Reply:
Yes that should do the trick!
I would probably move that logic to readPort and return 0 if the file is empty which would cause the loop to retry as currently there is a bit of a mix of concerns between the two functions.